src/test/cls_rgw/test_cls_rgw.cc

```diff
@@ -439,6 +441,72 @@ TEST(cls_rgw, index_list)
     ASSERT_EQ(it2->first.compare(keys[i]), 0);
 }
 
+
+TEST(cls_rgw, bi_list)
+{
+  string bucket_oid = str_int("bucket", 5);
+
+  OpMgr mgr;
+
+  ObjectWriteOperation *op = mgr.write_op();
+  cls_rgw_bucket_init(*op);
+  ASSERT_EQ(0, ioctx.operate(bucket_oid, op));
+
+  string name;
+  string marker;
+  int max = 10;
+  list<rgw_cls_bi_entry> entries;
+  bool is_truncated;
+
+  int ret = cls_rgw_bi_list(ioctx, bucket_oid, name, marker, max, &entries,
+			    &is_truncated);
+  ASSERT_EQ(ret, 0);
+  ASSERT_EQ(entries.size(), 0);
+  ASSERT_EQ(is_truncated, false);
+
+  uint64_t epoch = 1;
+
+  uint64_t obj_size = 1024;
+
+  int num_objs = 35;
+  for (int i = 0; i < num_objs; i++) {
+    string obj = str_int("obj", i);
+    string tag = str_int("tag", i);
+    string loc = str_int("loc", i);
+    index_prepare(mgr, ioctx, bucket_oid, CLS_RGW_OP_ADD, tag, obj, loc, RGW_BILOG_FLAG_VERSIONED_OP);
+    op = mgr.write_op();
+    rgw_bucket_dir_entry_meta meta;
+    meta.category = 0;
+    meta.size = obj_size;
+    index_complete(mgr, ioctx, bucket_oid, CLS_RGW_OP_ADD, tag, epoch, obj, meta, RGW_BILOG_FLAG_VERSIONED_OP);
+  }
+
+  ret = cls_rgw_bi_list(ioctx, bucket_oid, name, marker, num_objs + 10, &entries,
+			    &is_truncated);
+    ASSERT_EQ(ret, 0);
+    ASSERT_EQ(entries.size(), num_objs);
+
+  int num_entries = 0;
+
+  is_truncated = true;
+  while(is_truncated) {
+    ret = cls_rgw_bi_list(ioctx, bucket_oid, name, marker, max, &entries,
+			  &is_truncated);
+    ASSERT_EQ(ret, 0);
+    if (is_truncated) {
+      ASSERT_EQ(entries.size(), max);
```

Comment:
@oritwas not sure that we can assume that, backend might decide to return less entries than the max that we asked and still set trucated = true. We do want to check that we made some progress.

Reply:
@yehudasa , this way I found the off by one bug. If the backend changes the behavior we can modify the test.

Reply:
Is it 128k entries? Maybe 128?

Reply:
@yehudasa the default is 131072=128k

Reply:
@yehudasa, I guess it did change, I tried with 1000 and 2000 (total 3500 and 7000 entries) and it returned the requested max (I changed those values as the test execution time was too long). I don't recall the execution times, I will rerun an let you know. Maybe those are still acceptable, I prefer to use 1000 as we use in our code.

Reply:
possible that it was showing up in the rados suite? i see `suites/rados/basic/tasks/rgw_snaps.yaml:        osd_max_omap_entries_per_request: 10`

Reply:
I don't think so. Moreover, it seems to me that 128k per request is too high of a limit.

Reply:
it does seem pretty large, yeah

Reply:
since that rados/basic suite is running all of the cls tests with `osd_max_omap_entries_per_request: 10`, that should be sufficient coverage. but this test case would need to use different values of `int max = 10;` in order to show the difference

Reply:
:( long 1204942ms which are 20 min compared to 12617ms with the current test